Feature elastic indexing

Sallahuddin-kh opened the pull request:
This PR Accommodates the features such as:

- Django application creation
- Connection of Django application with Elasticsearch
- Index creation via Django
- Index population with Django commands

Sallahuddin-kh committed:
Adding skeleton project

Making index mappings
Making adding data to indexes

Sallahuddin-kh committed:
Refactoring code

atif-coeus commented on elasticassignment/elasticassignment/settings.py:
```diff
@@ -0,0 +1,141 @@
+
+DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'
+
+ELASTICSEARCH_DSL={
+    'default': {
+        'hosts': 'localhost:9200'
+    },
```
please use ENV vars.

atif-coeus commented on elasticassignment/ageofempires/views.py:
```diff
@@ -0,0 +1,13 @@
+
+
+def civilizations(request):
+    return render(request, 'ageofempires/civilizations.html' )
```
please always add new line at EOF

atif-coeus commented on elasticassignment/ageofempires/migrations/0001_initial.py:
```diff
@@ -0,0 +1,86 @@
+# Generated by Django 3.2.10 on 2021-12-17 07:55
```
why do we need migrations and DB?

Sallahuddin-kh committed:
removing migrations

atif-coeus commented on elasticassignment/ageofempires/models.py:
```diff
@@ -0,0 +1,69 @@
+from django.db import models
```
what will happen if someone runs `makemigration` command?

atif-coeus commented on elasticassignment/ageofempires/templates/ageofempires/civilizations.html:
```diff
@@ -0,0 +1,11 @@
+      <a href="https://www.freecodecamp.org/contribute/">The freeCodeCamp Contribution Page</a> shows you how and where you can contribute to freeCodeCamp's community and growth.
+    </p>
+  </body>
+</html>
```
new line?

atif-coeus commented on elasticassignment/elasticassignment/settings.py:
```diff
@@ -0,0 +1,143 @@
+
+DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'
+
+ELASTIC_HOST = 'localhost:9200'
+
+ELASTICSEARCH_DSL={
+    'default': {
+        'hosts': ELASTIC_HOST
+    },
+}
```
again, this is to be done via ENV vars.

If you don't have idea of ENV vars, please discuss with me.

atif-coeus commented on elasticassignment/elasticassignment/settings.py:
```diff
@@ -0,0 +1,143 @@
+# Database
+# https://docs.djangoproject.com/en/3.2/ref/settings/#databases
+
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
+        'NAME': BASE_DIR / 'db.sqlite3',
+    }
+}
```
unneeded config

atif-coeus commented on elasticassignment/ageofempires/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
would be good idea to add some tests!

atif-coeus replied:
tests are still missing!

atif-coeus commented on elasticassignment/ageofempires/documents.py:
```diff
@@ -0,0 +1,103 @@
+    class Index:
+        name = 'civilization'
+        settings = {
+            "number_of_shards": 1,
+            "number_of_replicas": 0
```
settings should be via ENV

atif-coeus commented on elasticassignment/ageofempires/management/commands/index_civilizations.py:
```diff
@@ -0,0 +1,45 @@
+from django.core.management.base import BaseCommand
```
Evaluate your design (of all `Command` classes) on SOLID principles: https://anuragdhadse.com/solid-principles/
Does it follow any of its principles?

atif-coeus commented on elasticassignment/ageofempires/management/commands/index_technologies.py:
```diff
@@ -0,0 +1,46 @@
+
+    def handle(self, *args, **kwargs):
+        config = {
+            'host': '127.0.0.1'
```
what is the purpose of `ELASTICSEARCH_DSL` setting your created in `settings.py` if you have to hard code the host here!!

Sallahuddin-kh committed:
Adding ENV vars

Making commands follow solid priciples
removinc unessecary settings

atif-coeus commented on elasticassignment/ageofempires/management/commands/command_classes.py:
```diff
@@ -0,0 +1,206 @@
+import elasticsearch
+from django.conf import settings
+
+class AOElasticsearch:
```
Please create a package `Elasticsearch` with a module `client`.

atif-coeus commented on elasticassignment/ageofempires/management/commands/command_classes.py:
```diff
@@ -0,0 +1,206 @@
+        pass
+
+
+class Civilizations(AOElasticsearch):
```
please create a package `models` and put all model classes inside in their own files.

atif-coeus commented on elasticassignment/ageofempires/management/commands/command_classes.py:
```diff
@@ -0,0 +1,206 @@
+    def make_connection():
+        return elasticsearch.Elasticsearch([settings.CONFIG])
+
+    def getAPIdata(url:str):
```
`get_data`

atif-coeus commented on elasticassignment/ageofempires/management/commands/command_classes.py:
```diff
@@ -0,0 +1,206 @@
+
+    def make_index():
+        request_body = {
+            "settings" : {
+                "index":{
+                    "number_of_shards": settings.NUMBER_OF_SHARDS,
+                    "number_of_replicas": settings.NUMBER_OF_REPLICAS
+                }
+            },
```
please create a `__get_settings` method.
and also a `__get_mappings` method.

We'll further refactor this part in next review cycle.

atif-coeus commented on elasticassignment/ageofempires/management/commands/command_classes.py:
```diff
@@ -0,0 +1,206 @@
+                "_source": structure
+            }
+            structures_list.append(struc)
+        bulk(Structures.es, structures_list, 20)
```
as explained, please don't use the batch size param in `bulk` function in this particular exercise.
The batch size is to be controlled by your python program.

atif-coeus commented on elasticassignment/ageofempires/management/commands/index_technologies.py:
```diff
@@ -1,46 +1,8 @@
-        bulk(connections.get_connection(),technologies_list,20)
+    def handle(self, *args, **kwargs):        
+        command_classes.Technologies.make_index()
+        command_classes.Technologies.populate_index("https://age-of-empires-2-api.herokuapp.com/api/v1/technologies")
```
Please get the BASE URL from ENV.

eg. AOE_BASE_URL = 'https://age-of-empires-2-api.herokuapp.com'

atif-coeus replied:
please create a setting `AOE_ENDPOINTS` and put all paths in that setting.

atif-coeus commented on elasticassignment/elasticassignment/settings.py:
```diff
@@ -0,0 +1,136 @@
+
+# Initialise environment variables
+env = environ.Env()
+environ.Env.read_env()
```
please add `.env.example` file

Sallahuddin-kh committed:
Adding composition based structure

Adding urls as environment vars and settings
Controlling batch sizes from python code

Sallahuddin-kh committed:
Adding setting AOE_ENDPOINTS

Sallahuddin-kh committed:
Fixing indentation

atif-coeus commented on elasticassignment/ageofempires/models/structures.py:
```diff
@@ -0,0 +1,61 @@
+from django.conf import settings
+from ageofempires.Elastic_search.elastic_client import elastic_client
+
+class Structures():
```
Please always use singular for model names

atif-coeus commented on elasticassignment/ageofempires/Elastic_search/elastic_client.py:
```diff
@@ -0,0 +1,12 @@
+import elasticsearch
+from django.conf import settings
+
+class elastic_client:
```
- correct the name: `ElasticClient`
- Please make `make_connection` method private.
- add `__init__` function and instantiate `self.__connection`.
- correct the functionality of `get_data` function. It should retrieve data from Elastic indexes.
- add `index_data` function.
- add `create_index` function, it should accept `settings` and `mappings` as params.
- add any other function that needs direct connection with Elasticsearch.

atif-coeus commented on elasticassignment/ageofempires/Elastic_search/elastic_client.py:
```diff
@@ -0,0 +1,12 @@
+    def make_connection():
+        return elasticsearch.Elasticsearch([settings.CONFIG])
+
+    def get_data(url:str):
+        response_data = requests.get(url)
+        return response_data.json()
```
The functionality in this function is not related to Elastic Client. 

For HTTP calls please create a separate client class `HTTPClient` with `get` method.

atif-coeus commented on elasticassignment/ageofempires/management/commands/index_civilizations.py:
```diff
@@ -0,0 +1,10 @@
+import ageofempires.models.civilizations as Civilizations
+from django.conf import settings
+
+class Command(BaseCommand):
```
- create single command `import_data` for all imports and this command should accept `import_name` param.
- keep the `import_name` param optional. If not given, all imports should execute.

atif-coeus replied:
- Create a command `create_indexes` which should create all indexes for the app.

atif-coeus replied:
- In `Elastic_search` package, please add a module `index_config`.
- add `get_index_settings` method.
- add `get_index_mappings` method which should accept `index_name` param.

atif-coeus commented on elasticassignment/ageofempires/models/civilizations.py:
```diff
@@ -0,0 +1,52 @@
+        es  = elastic_client.make_connection()
+        es.indices.create(index = 'civilization', body = request_body)
+
+    def populate_index(url:str):
```
- please redefine as: `bulk_index(data:list)`
- move data processing and batching out of this function (will move to the `import_data` command)

atif-coeus replied:
- model classes will also need `get_objects` or similar function.

Sallahuddin-kh committed:
Adding elastic client module and refactoring

Sallahuddin-kh committed:
Fixing indentation

Sallahuddin-kh committed:
Adding UI for indexes

Sallahuddin-kh committed:
Adding code docs and type hints

atif-coeus commented on elasticassignment/ageofempires/ElasticClient/Client.py:
```diff
@@ -0,0 +1,27 @@
+from django.conf import settings
+from elasticsearch.helpers import bulk
+
+class Client:
```
Now it looks better 👍 

How do you think about adding some codedocs?

atif-coeus replied:
Please add codeblocks in all classes added by you (not by the framework)

atif-coeus commented on elasticassignment/ageofempires/models/unit.py:
```diff
@@ -0,0 +1,18 @@
+        self.client.create_index('units',settings,mappings)
+
+    def get_objects(self):
+        return self.client.get_data('units')
```
new line?

atif-coeus commented on elasticassignment/ageofempires/models/unit.py:
```diff
@@ -0,0 +1,18 @@
+    def make_index(self):
+        settings = conf.get_index_settings()
+        mappings = conf.get_index_mappings('units')
+        self.client.create_index('units',settings,mappings)
```
missing space after commas

atif-coeus commented on elasticassignment/ageofempires/management/commands/create_indexes.py:
```diff
@@ -0,0 +1,17 @@
+class Command(BaseCommand):
+
+    def handle(self, *args, **kwargs):
+        civ_intance = Civilization.Civilization()
+        struc_instance = Structure.Structure()
+        tech_instance = Technology.Technology()
+        uni_instance = Unit.Unit()
```
- spelling mistake in `civ_intance`
- instead of using shorts, its better to use full names for objects. For example, `civilization` instead of `civ_instance`.
- To call constructor you don't need to do `ClassName.Classname()`. `ClassName()` works well.

atif-coeus replied:
Throughout this PR, please check your variable naming and correct them:
> instead of using shorts, its better to use full names for objects. For example, civilization instead of civ_instance.

atif-coeus commented on elasticassignment/ageofempires/models/technology.py:
```diff
@@ -0,0 +1,18 @@
+        self.client.create_index('technologies',settings,mappings)
+
+    def get_objects(self):
+        return self.client.get_data('technologies')
```
new line missing!!

atif-coeus commented on elasticassignment/ageofempires/models/unit.py:
```diff
@@ -0,0 +1,18 @@
+import ageofempires.ElasticClient.Client as Client
```
If you check all your model classes, there is repeated code which means all the model classes are performing same kind of operations (which is expected from them)!

However, this kind of similar classes give us the opportunity to implement a basic OOP concept. What's that and how will you implement it? (Please discuss before making changes)

Sallahuddin-kh committed:
refactoring and adding testcases

Sallahuddin-kh committed:
Fixing copy paste mistake in views.py

Sallahuddin-kh committed:
Merge pull request #2 from Sallahuddin-kh/feature/app-UI

Feature/app UI
